### ISSUE TITLE
Ctrl+D kills the REPL too easily

### ISSUE THREAD

**Issue by lgrapenthin** [bhauman/rebel-readline#77]
Love trying out this software so far. However I always terminate it by accidentally typing CTRL+D.
Probably it would help to ask for confirmation before quitting the REPL.

**Comment by bhauman** [comment]
Ctrl-D should only kill the REPL at the start of the line. Is this the case for you?

**Comment by lgrapenthin** [comment]
Yes. Given that many shortcuts start with C-x its quite easy to accidentally hit C-d, 'd' being very close to 'x'....

**Comment by bhauman** [comment]
cool just verifying, this is built in behavior of the current Clojure REPL. Adn builltIn Jline# behavior. I'll look into just using :repl/quit. As I think that is the preferred solution to this.

**Comment by trptcolin** [comment]
For what it’s worth, quitting on ctrl-d is also the behavior in readline stuff for bash/zsh, ruby, python, node, haskell, mysql, postgresql, and others. Erlang/Elixir is the only readline I’m aware of where this is not the case (I assume there are others?). So I think this could be surprising for folks. Maybe OK - there’s lots of convenient but unconventional stuff here - but worth considering I think. Searching for `ctrl-d erlang` shows some of the confusion in that world (which I encountered when getting started over there).

So I would personally be :-1: on this change as a default, but if there’s a way to toggle it back on, I’m less worried about it.

**Comment by bhauman** [comment]
great input Colin! thanks!

**Comment by lgrapenthin** [comment]
I recommended a confirmation prompt because it wouldn't be confusing and prevent accidental quits.

**Comment by trptcolin** [comment]
Another minor thing to consider that doesn't affect the core functionality but could be a bigger philosophical question: with a different way to quit beyond ctrl-d/EOF, piping STDIN into the process requires some additional information to be composed on. Not the end of the world, but a program producing the input would need to know more details of how rebel works internally. For example:
```
$ echo '(println "hello")' | REBEL_COMMANDLINE_HERE
```
would need to become something along the lines of:
```
$ echo '(println "hello")\ny\n' | REBEL_COMMANDLINE_HERE
```
or 
```
$ echo '(println "hello")\n:repl/quit\n' | REBEL_COMMANDLINE_HERE
```
or similar. Sending text in from outside the process isn't the typical workflow of course, and the right thing to do here is really to pass to a simpler program like the raw `clojure.main`, but we've needed to consider it in the past for leiningen/reply so thought I'd mention it.

**Comment by bhauman** [comment]
This is a very interesting point, as well. Again thanks!

**Comment by christoph-frick** [comment]
A single `CTRL-D` should quit at once (no questions asked).  This is the way pretty much anything REPL-y/Shell-y works (in Unix). 👎 here too

**Comment by lgrapenthin** [comment]
Good points, agreed. Probably I am the only user who fails at not terminating.

Otherwise, a beginner friendly flag/param to main would be nice.

**Comment by trptcolin** [comment]
Doing some Python for the first time this week, and I noticed that ipython has a Y/n prompt after hitting ctrl-d. Another ctrl-d does the trick at this second prompt too. So I wanted to bring that up to argue against myself a bit. It hasn't felt like a major annoyance having to hit ctrl-d twice, and this also works when piping text into ipython.

**Comment by christoph-frick** [comment]
Maybe another thing to consider: ctlr-d on its own closes the repl/shell at once,  but e.g. in python (vi) it executes the command first that is entered.  zsh (vi) and psql (non-vi) both ignore it complelty unless it's the first thing in a line. rebel-rl currently behaves like zsh/psql in that matter (vi) -- i added my mode in (), but i assume from the psql behaviour, that it does not matter.

**Comment by borkdude** [comment]
In addition, ctrl-z puts the process in the background. Maybe ctrl-z and/or CMD-Z should be undo in rebel-readline.

**Comment by christoph-frick** [comment]
This is also common behaviour.  `fg` will bring you back into the program.  Undo works as with your editor key mapping.  E.g. it's `u` with the `:vi` modes (it does not work properly, but meh).  If ctlr-z IS undo in emacs, i would be curious, what readline/rlwrap has picked for it.
